{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e7633580",
   "metadata": {},
   "source": [
    "Exercise II. Genomic Signatures\n",
    "Task 1: Obtain the first chromosome from the genomes of the following organisms: (H. sapiens, M. musculus, R. norvegicus, D. melanogaster, C. elegans and S. cerevisiae). The sequences need to be in fasta format or similar. You may use the database of your choice but you are advised to opt for the UCSC Table Browser.\n",
    "Task 2: Calculate a table of o/e 2-mer frequencies for a given sequence. Use mononucleotide frequencies as background.\n",
    "Task 3: Calculate the Manhattan distance of the k-mer frequency for all sequence pairs\n",
    "Task 4: Return a table with the Manhattan distances alongside your code and a brief discussion of your results. Does the Distance table correspond to what you would expect in terms of distances between the species?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80066d91",
   "metadata": {},
   "source": [
    "To test the following code all you have to do is download some genomes files in fasta format and save them with a relative path \"./data/*.fasta\"\n",
    "The output is going to be a table with the 2-mers frequencies for each given genome, and a table with Manhattan distances between the choosen genomes you use as input. \n",
    "Note, that you can change the k and use the same code for any k-mers, you want, but maybe it will take some more time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "87d75948",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                 CA        AG        GT        TA        AA  \\\n",
      "mouse_chr1.fasta           1.234818  1.214728  0.884626  0.750466  1.059793   \n",
      "c.elegan_chrI.fasta        1.064757  0.896344  0.823121  0.603778  1.303311   \n",
      "hg39_chr1.fasta            1.194468  1.097827  0.798628  0.707548  1.125449   \n",
      "d.melanogaster_chr2.fasta  1.126019  0.890284  0.853149  0.752908  1.218975   \n",
      "rat_chr1.fasta             1.229868  1.186669  0.880994  0.737624  1.073928   \n",
      "s.cerevisiae_chrI.fasta    1.124746  0.981633  0.930224  0.762388  1.130446   \n",
      "\n",
      "                                 GA        AC        AT        TT        GG  \\\n",
      "mouse_chr1.fasta           1.040431  0.885620  0.863618  1.062065  1.192180   \n",
      "c.elegan_chrI.fasta        1.115394  0.822504  0.857755  1.302336  1.060995   \n",
      "hg39_chr1.fasta            1.034884  0.802226  0.940917  1.099674  1.254045   \n",
      "d.melanogaster_chr2.fasta  0.912362  0.855301  0.963295  1.219659  1.055138   \n",
      "rat_chr1.fasta             1.042834  0.877588  0.877687  1.065489  1.208961   \n",
      "s.cerevisiae_chrI.fasta    1.042571  0.922408  0.931326  1.135692  1.036710   \n",
      "\n",
      "                                 CT        TG        GC        TC        CC  \\\n",
      "mouse_chr1.fasta           1.212928  1.237136  0.931572  1.030763  1.194218   \n",
      "c.elegan_chrI.fasta        0.894744  1.062462  1.051592  1.113697  1.063488   \n",
      "hg39_chr1.fasta            1.150649  1.217842  0.973446  1.059502  1.228789   \n",
      "d.melanogaster_chr2.fasta  0.892370  1.126794  1.271499  0.912228  1.052510   \n",
      "rat_chr1.fasta             1.190150  1.232321  0.899332  1.048131  1.207548   \n",
      "s.cerevisiae_chrI.fasta    0.966270  1.123027  1.005029  1.032894  1.064552   \n",
      "\n",
      "                                 CG  \n",
      "mouse_chr1.fasta           0.179849  \n",
      "c.elegan_chrI.fasta        1.019764  \n",
      "hg39_chr1.fasta            0.293426  \n",
      "d.melanogaster_chr2.fasta  0.921469  \n",
      "rat_chr1.fasta             0.192565  \n",
      "s.cerevisiae_chrI.fasta    0.798254  \n",
      "                           mouse_chr1.fasta  c.elegan_chrI.fasta  \\\n",
      "mouse_chr1.fasta                   0.000000             3.122010   \n",
      "c.elegan_chrI.fasta                3.122010             0.000000   \n",
      "hg39_chr1.fasta                    0.917868             2.652247   \n",
      "d.melanogaster_chr2.fasta          2.951735             1.358070   \n",
      "rat_chr1.fasta                     0.211573             3.063276   \n",
      "s.cerevisiae_chrI.fasta            1.991499             1.503113   \n",
      "\n",
      "                           hg39_chr1.fasta  d.melanogaster_chr2.fasta  \\\n",
      "mouse_chr1.fasta                  0.917868                   2.951735   \n",
      "c.elegan_chrI.fasta               2.652247                   1.358070   \n",
      "hg39_chr1.fasta                   0.000000                   2.585243   \n",
      "d.melanogaster_chr2.fasta         2.585243                   0.000000   \n",
      "rat_chr1.fasta                    0.775584                   2.930091   \n",
      "s.cerevisiae_chrI.fasta           1.774615                   1.199446   \n",
      "\n",
      "                           rat_chr1.fasta  s.cerevisiae_chrI.fasta  \n",
      "mouse_chr1.fasta                 0.211573                 1.991499  \n",
      "c.elegan_chrI.fasta              3.063276                 1.503113  \n",
      "hg39_chr1.fasta                  0.775584                 1.774615  \n",
      "d.melanogaster_chr2.fasta        2.930091                 1.199446  \n",
      "rat_chr1.fasta                   0.000000                 1.984641  \n",
      "s.cerevisiae_chrI.fasta          1.984641                 0.000000  \n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import glob\n",
    "import os\n",
    "import pandas as pd\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "\n",
    "#creating a dictionary with all the kmers in a given sequence\n",
    "def find_kmers(seq,k):\n",
    "    kmers = {}\n",
    "    for i in range(len(seq)-k +1):\n",
    "        subseq = seq[i:i+k]\n",
    "        if subseq not in kmers.keys():\n",
    "            kmers[subseq] = 1\n",
    "        else:\n",
    "            kmers[subseq] += 1\n",
    "    return kmers\n",
    "\n",
    "#calculate the background frequencies, by calculate the  mononucleotide frequencies\n",
    "def background_freq(seq):\n",
    "    seq = seq.upper()\n",
    "    total = len(seq)\n",
    "    freq_dict = {\n",
    "        'A': seq.count('A') / total,\n",
    "        'C': seq.count('C') / total,\n",
    "        'G': seq.count('G') / total,\n",
    "        'T': seq.count('T') / total\n",
    "    }\n",
    "    return freq_dict\n",
    "\n",
    "def parse_fasta(filepath):\n",
    "    file_kmer_data = {}\n",
    "\n",
    "#reading all the files I want to use \n",
    "    for file in glob.glob(filepath):\n",
    "        file_name = os.path.basename(file)\n",
    "        all_kmers = {}\n",
    "        count = 0\n",
    "\n",
    "#parsing each file\n",
    "        for record in SeqIO.parse(file, \"fasta\"):\n",
    "            sequence = ''.join([base for base in str(record.seq).upper() if base in 'ACGT'])\n",
    "#calling the first function to find the 2-mers in every sequence\n",
    "            kmers = find_kmers(sequence,2)\n",
    "#calculate the observe k-mers freq, by divide with the lenght of the whole seq for normalization\n",
    "            kmers_freq = {kmer: count / len(sequence) for kmer, count in kmers.items()}\n",
    "#calling the second function to calculate mononucleotide frequencies for each seq (expected frequencies)        \n",
    "            freq_seq = background_freq(sequence)\n",
    "\n",
    "#calculating o/e frequencies for each k-mer (observe freq/(freq of the first base)*(freq of the second base))            \n",
    "            normalized_kmers_freq = {\n",
    "                kmer: freq / (freq_seq.get(kmer[0], 1e-6)* freq_seq.get(kmer[1], 1e-6))\\\n",
    "                for kmer, freq in kmers_freq.items()\n",
    "            }\n",
    "#suming all the o/e freq we calculate for every k-mer in a final dict for each file, \n",
    "# this step is needed because we didn't parse the files manually  \n",
    "            for kmer, val in normalized_kmers_freq.items():\n",
    "                all_kmers.setdefault(kmer, []).append(val)\n",
    "            count += 1\n",
    "#we calculating the mean of the freq for every k-mer, to estimate a total freq for every file\n",
    "        file_kmer_data[file_name] = {k: sum(v) / len(v) for k, v in all_kmers.items()}\n",
    "\n",
    "#transform the final_dict to a DataFrame\n",
    "        df = pd.DataFrame.from_dict(file_kmer_data, orient='index').fillna(0)\n",
    "    \n",
    "    return df\n",
    "\n",
    "#You can use the correct relative path, for the fasta files in your directory to use the code\n",
    "df = parse_fasta(\"./data/*.fasta\")\n",
    "\n",
    "print(df)\n",
    "\n",
    "#calculating Manhattan distance and save the output in a new DataFrame\n",
    "distance_matrix = pdist(df.values, metric='cityblock')\n",
    "\n",
    "distance_df = pd.DataFrame(squareform(distance_matrix), index=df.index, columns=df.index)\n",
    "\n",
    "print(distance_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcfda2e0",
   "metadata": {},
   "source": [
    "RESULTS:\n",
    "\n",
    "        mouse_chr1.fasta  c.elegan_chrI.fasta  \\\n",
    "mouse_chr1.fasta                   0.000000             3.122010   \n",
    "c.elegan_chrI.fasta                3.122010             0.000000   \n",
    "hg39_chr1.fasta                    0.917868             2.652247   \n",
    "d.melanogaster_chr2.fasta          2.951735             1.358070   \n",
    "rat_chr1.fasta                     0.211573             3.063276   \n",
    "s.cerevisiae_chrI.fasta            1.991499             1.503113   \n",
    "\n",
    "                           hg39_chr1.fasta  d.melanogaster_chr2.fasta  \\\n",
    "mouse_chr1.fasta                  0.917868                   2.951735   \n",
    "c.elegan_chrI.fasta               2.652247                   1.358070   \n",
    "hg39_chr1.fasta                   0.000000                   2.585243   \n",
    "d.melanogaster_chr2.fasta         2.585243                   0.000000   \n",
    "rat_chr1.fasta                    0.775584                   2.930091   \n",
    "s.cerevisiae_chrI.fasta           1.774615                   1.199446   \n",
    "                          rat_chr1.fasta  s.cerevisiae_chrI.fasta  \n",
    "mouse_chr1.fasta                 0.211573                 1.991499  \n",
    "c.elegan_chrI.fasta              3.063276                 1.503113  \n",
    "hg39_chr1.fasta                  0.775584                 1.774615  \n",
    "d.melanogaster_chr2.fasta        2.930091                 1.199446  \n",
    "rat_chr1.fasta                   0.000000                 1.984641  \n",
    "s.cerevisiae_chrI.fasta          1.984641                 0.000000  \n",
    "\n",
    "\n",
    "As we can see from the final table, we conclude that organisms known to be closely related in evolutionary terms, such as rats and mice, have a small Manhattan distance (0.21). In contrast, organisms with a greater evolutionary distance, such as mice and C. elegans, show a much larger distance (3.1). From this, we can infer that biologically related organisms tend to have more similar genome sequences — reflected in smaller genetic distances — and also tend to maintain similar 2-mer frequency patterns.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
